### PR TITLE
doc/rgw: more detail on transitions between log generations

### PR DESCRIPTION
**Pull request by cbodley**
updates from further design discussions

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/doc/rgw/multisite-reshard.md [hunk 1]
```diff
@@ -95,9 +105,14 @@ Reframe in terms of log generations, instead of handling SYNCSTOP events with a
 
 ### Admin APIs
 
-* RGWOp_BILog_List response should include the bucket's highest log generation
+* RGWOp_BILog_List
+    - Add query parameter for 'generation' that defaults to the latest generation.
+    - Add query parameter for 'format-ver' that defaults to 1. If format-ver >= 2, return the results as a struct that includes:
+        * A truncated flag that indicates whether more entries are available.
+        * An optional 'next_generation' field that includes 'num_shards' and generation number of the next log generation.
```
Comment (smanjara) [inline src/doc/rgw/multisite-reshard.md:112]: A struct 'next_gen' that includes num_shards and gen?

Comment (cbodley) [reply src/doc/rgw/multisite-reshard.md:112]: that's the info we'll need for the transition, yeah

File: src/doc/rgw/multisite-reshard.md [hunk 2]
```diff
@@ -53,7 +53,10 @@ The distinction between *index layout* and *log layout* is important, because in
 
 ### Data Sync
 
-* Datalog entries currently include a bucket shard number. We need to add the log generation number to these entries so we can tell which sharding layout it refers to. If we see a new generation number, that entry also implies an obligation to finish syncing all shards of prior generations.
+* Add the log generation number to datalog entries so we can tell which log layout they refer to.
+* If we see a new generation number in the datalog, that entry also implies an obligation to finish syncing all shards of prior generations:
+    - For each shard of the current generation that's not in the 'done' state, write an error_repo entry to its corresponding datalog shard.
+    - Write the initial datalog entry to the error_repo so it's retried until we reach that generation.
```
Comment (adamemerson) [inline src/doc/rgw/multisite-reshard.md:59]: Thank you1 This is the detail I wasn't sure about.

**Review by adamemerson — APPROVED** [review]

**Comment by smanjara** [comment]
unstale

File: src/doc/rgw/multisite-reshard.md [hunk 3]
```diff
@@ -66,15 +69,22 @@ The distinction between *index layout* and *log layout* is important, because in
...
 * Full sync uses a single bucket-wide listing to fetch all objects.
     - Use a cls_lock to prevent different shards from duplicating this work.
 * When incremental sync gets to the end of a log shard (i.e. listing the log returns truncated=false):
     - If the remote has a newer log generation, flag that shard as 'done' in the bucket sync status.
-    - Once all shards in the current generation reach that 'done' state, incremental bucket sync can advance to the next generation.
-    - Use cls_version on the bucket sync status object to detect racing writes from other shards.
+    - Once all shards in the current generation reach that 'done' state, incremental bucket sync can transition to the next generation.
+* On transition:
+    - Clear the set of 'done' shards and update the num_shards and generation in the bucket sync status.
+    - If num_shards was reduced, delete any extra per-shard sync status objects.
```
Comment (smanjara) [inline src/doc/rgw/multisite-reshard.md:87]: @cbodley sorry I had missed looking at this pr earlier. what does extra status object mean? once a shard has completed syncing for a generation, we just have to delete that shard status object right?

**Pull request closed without merging**